src/tactic/norm_num.lean

```diff
@@ -1206,6 +1206,20 @@ theorem dvd_eq_nat (a b c : ℕ) (p) (h₁ : b % a = c) (h₂ : (c = 0) = p) : (
 theorem dvd_eq_int (a b c : ℤ) (p) (h₁ : b % a = c) (h₂ : (c = 0) = p) : (a ∣ b) = p :=
 (propext $ by rw [← h₁, int.dvd_iff_mod_eq_zero]).trans h₂
 
+theorem int_to_nat_pos (a : ℤ) (b : ℕ) (h : (by haveI := @nat.cast_coe ℤ; exact b : ℤ) = a) :
```

Comment:
I guess the `haveI` in `h` is to make sure the statement uses a certain `has_coe` instance? Why is this needed? Could you add a comment explaining?

Reply:
It's just an internal implementation detail - `prove_nat_uncast` uses `nat.cast` on any type, including `nat -> int`, but in this case lean prefers a different (not defeq) instance coming from `int.of_nat`. The statement of that particular hypothesis is intended to match `prove_nat_uncast`'s conclusion. It would be an option to write a copy of `prove_nat_uncast` that uses `int.of_nat` but both proofs are the same length, you just apply this lemma and then the relevant uncast lemmas.

Reply:
I'm not sure I want to set a precedent of explaining these hypotheses - all the lemmas in this file are kind of weird, like `dvd_eq_int` just above it which uses a `(c = 0) = p` hypothesis (not to mention equality hypotheses in general, and too many explicit arguments). In all cases the answer for why they are written that way is the same - see the part of the code that uses the lemma, it's set up to exactly match (no defeq reduction allowed) the conclusion of some lemma generated by a proof procedure, which should describe the shape of the lemma in its docstring.

Reply:
I think what you've just written would be useful to put in a module doc string somewhere in the file.

Reply:
@digama0 How about adding this around line 58:
```lean

/-!
Each lemma in this file is written the way it is to exactly match (with no defeq reduction allowed)
the conclusion of some lemma generated by the proof procedure that uses it. That proof procedure
should describe the shape of the generated lemma in its docstring.
-/
```